tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs

```diff
@@ -0,0 +1,147 @@
+                            serviceName.Should().Be("mongodb"); // The default is "mongodb".
+                            var canonicalizeHostName = authenticator._mechanism()._canonicalizeHostName();
+                            canonicalizeHostName.Should().BeFalse(); // The default is "false".
+                        }
+                        else
+                        {
+                            // Other authenticators do not contain mechanism properties
+                        }
+                    }
+                    else
+                    {
+                        foreach (var expectedMechanismProperty in expectedMechanismProperties.AsBsonDocument)
+                        {
+                            var mechanismName = expectedMechanismProperty.Name;
+                            switch (mechanismName)
+                            {
+                                case "SERVICE_NAME":
+                                    var serviceName = authenticator._mechanism()._serviceName();
+                                    serviceName.Should().Be(ValueToString(expectedMechanismProperty.Value));
+                                    break;
+                                case "CANONICALIZE_HOST_NAME":
+                                    var canonicalizeHostName = authenticator._mechanism()._canonicalizeHostName();
+                                    canonicalizeHostName.Should().Be(expectedMechanismProperty.Value.ToBoolean());
+                                    break;
+                                default:
+                                    throw new Exception($"Invalid mechanism property '{mechanismName}'.");
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        private void AssertInvalid(Exception ex, BsonDocument definition)
+        {
+            if (definition["valid"].ToBoolean())
+            {
+                throw new AssertionException($"The connection string '{definition["uri"]}' should be valid.", ex);
+            }
+        }
+
+        private string ValueToString(BsonValue value)
+        {
+            return value == BsonNull.Value ? null : value.ToString();
+        }
+
+        // nested types
+        private class TestCaseFactory : JsonDrivenTestCaseFactory
+        {
+            protected override string PathPrefix => "MongoDB.Driver.Tests.Specifications.auth.tests.";
+        }
+    }
+
+    internal static class AuthenticatorReflector
+    {
+        public static object _mechanism(this IAuthenticator obj) => Reflector.GetFieldValue(obj, nameof(_mechanism));
+    }
+
+    internal static class GssapiMechanismReflector
+    {
+        public static bool _canonicalizeHostName(this object obj) => (bool)Reflector.GetFieldValue(obj, nameof(_canonicalizeHostName));
+        public static string _serviceName(this object obj) => (string)Reflector.GetFieldValue(obj, nameof(_serviceName));
```

Comment:
We really should not be defining an extension method on `object`.

Find the right type and reflect on that.

Reply:
`GssapiMechanism` is nested private class, so I cannot do much here, apart from moving the retrieval of `_canonicalizeHostName` and `_serviceName` into `AuthTestRunner`, but that would violate the existing style of using reflections in unit tests.

Reply:
This reflector can be eliminated by implementing the previous comment.